core/modal_models/modes.h

```diff
@@ -107,9 +110,9 @@ void _lf_handle_mode_changes(void);
 void _lf_handle_mode_triggered_reactions(void);
 
 /**
- * Function (to be code generated) to get the reactor_mode_state array and its length
+ * Function (to be code generated) to fill up a vector with the reactors who has mode transitions
```

Comment:
```suggestion
 * Function (to be code generated) to fill up a vector with the reactors which have mode transitions.
```
Also, is there no restriction to nested reactors?

Reply:
I am not entirely sure what you mean? I believe that reactors contained within the mode also should be added to this vector. I will make a test for it. Will clarify the comment

Reply:
Good catch. I had misunderstood the implementation. It is indeed right that we dont add reactors **contained** in the mode. I added a failing test, will address it.

Reply:
Is there a way of seeing if a Reactor is local to a specific mode? Worst-case is that we have to lock ALL contained reactors in any reactor which is gonna do a mode change.  Ideally we only lock those that are contained in the mode we are transitioning away from

Reply:
Actually. Is it even a way of getting the containment hierarchy in the run-time? Might have to code-generate a function for this?

Reply:
No, at the moment you cannot see if a Reactor is local to a mode. I was a minimalist in introducing pointers for these structures and designed them to be unidirectional and upward in containment, to circumvent having to allocate collections of references.
Reactions are associated with their directly enclosing mode, because these need to suppressed upon inactivity. Reactors only have a reference to an enclosing mode if they are modal reactors, in order to establish a mode hierarchy.
Hence, you can get the enclosing mode via any contained reaction or the mode state; or you need to introduce another pointer if that is not sufficient.

As far as I know the regular containment relations between reactors are not available at runtime, otherwise I would have utilized them for modes, I guess.

Reply:
Had a brief discussion with @edwardalee yesterday where he proposed that we add the containment hierarchy to the self_struct by adding a `self_base_t *parent` pointer. We then do the following:

1. Before reaction invocation the worker will acquire the local mutex of the reactor but also traverse up the containment hierarchy and acquire the local mutex of any modal reactor up its containment chain. 
2. Before advancing time the worker will acquire the local mutex of any modal reactor which is about to perform a mode change. If this modal reactor has a contained reactor with a LET reaction it will have to wait on that mutex.

There are three features of this design
1. We don't have to "propagate" down the modal hierarchy and lock the contained modal reactors as well **before advancing time**. It will be sufficient to lock the modal reactors which have requested mode change.
2. It will be conservative as a reactor OUTSIDE any mode will also block mode transition. 
3. We can no longer execute LET reactions in independent reactors contained in a modal reactor in parallel. If e.g. a mode has 2 contained reactors, both with LET reactions, both will try to acquire the local mutex of the parent modal reactor. Marten suggested that a semaphore might solve this